jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/QueuedPool.java

```diff
@@ -0,0 +1,245 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Stream;
+
+/**
+ * This implementation does not support multiplexing.
+ *
+ * @param <P> the type of the pooled objects
+ */
+public class QueuedPool<P> implements Pool<P>
+{
+    private final int maxSize;
+    private final Queue<Pool.Entry<P>> queue = new ConcurrentLinkedQueue<>();
+
+    // This lock protects the 'terminated' field.
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    private final Lock rdLock = rwLock.readLock();
+    private final Lock wrLock = rwLock.writeLock();
+    private boolean terminated;
+
+    public QueuedPool(int maxSize)
+    {
+        this.maxSize = maxSize;
+    }
+
+    @Override
+    public Entry<P> reserve()
+    {
+        rdLock.lock();
+        try
+        {
+            if (terminated || queue.size() == maxSize)
+                return null;
+            return new QueuedEntry<>(this);
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    private boolean requeue(Pool.Entry<P> entry)
+    {
+        rdLock.lock();
+        try
+        {
+            if (terminated || queue.size() == maxSize)
+                return false;
+            queue.add(entry);
+            return true;
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    @Override
+    public Entry<P> acquire()
+    {
+        rdLock.lock();
+        try
+        {
+            if (terminated)
+                return null;
+            QueuedEntry<P> entry = (QueuedEntry<P>)queue.poll();
+            if (entry != null)
+                entry.acquire();
+            return entry;
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    @Override
+    public boolean isTerminated()
+    {
+        rdLock.lock();
+        try
+        {
+            return terminated;
+        }
+        finally
+        {
+            rdLock.unlock();
+        }
+    }
+
+    @Override
+    public Collection<Entry<P>> terminate()
+    {
+        wrLock.lock();
+        try
+        {
+            terminated = true;
+            Collection<Entry<P>> copy = new ArrayList<>(queue);
+            queue.clear();
+            return copy;
+        }
+        finally
+        {
+            wrLock.unlock();
+        }
+    }
+
+    @Override
+    public int size()
+    {
+        return queue.size();
+    }
+
+    @Override
+    public int getMaxSize()
+    {
+        return maxSize;
+    }
+
+    @Override
+    public Stream<Entry<P>> stream()
+    {
+        return queue.stream();
+    }
+
+    private static class QueuedEntry<P> implements Entry<P>
+    {
+        private final QueuedPool<P> pool;
+        //   null  -> terminated
+        //   false -> idle
+        //   true  -> in use
+        private volatile Boolean inUse = Boolean.FALSE;
+        private volatile P pooled;
```

Comment:
Two volatile fields just screams race to me!   I preferred my version is an AtomicMarkableReference

Reply:
It's not racy, but I agree it's ugly. We need four states: reserved, enabled/available, enabled/in-use, terminated but that can be encoded in a `AtomicMarkableReference` so I'm going to use that instead as you recommend.